lisa/tests/hotplug/rollback.py

```diff
@@ -0,0 +1,195 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (C) 2021, Arm Limited and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import pandas as pd
+
+from devlib.module.hotplug import HotplugModule
+from devlib.exception import TargetStableError
+
+from lisa.tests.base import ResultBundle, TestBundle
+from lisa.trace import FtraceCollector, Trace
+from lisa.utils import ArtifactPath
+
+
+class HotplugRollback(TestBundle):
+
+    TRACE_PATH = 'trace.dat'
+
+    @classmethod
+    def _online(cls, target, cpu, online, verify=True):
+        try:
+            if online:
+                target.hotplug.online(cpu)
+            else:
+                target.hotplug.offline(cpu)
+        except Exception as e:
+            if verify:
+                raise e
+
+    @classmethod
+    def _state_can_fail(cls, target, cpu, state, up):
+        """
+        There are no way of probing the kernel for a list of hotplug states that
+        can fail and for which we can test the rollback. We need therefore to
+        try:
+          - If we can set the state in the kernel 'fail' interface.
+          - If the hotplug is reset actually failing (some states can fail only
+            when going up or down)
+        """
+        try:
+            target.hotplug.fail(cpu, state)
+            try:
+                cls._online(target, cpu, up)
+                target.hotplug.fail(cpu, -1)
+                cls._online(target, cpu, not up)
+                return False
+            except TargetStableError:
+                return True
+        except TargetStableError:
```

Comment:
looks like this could be flattened into:
```
try:
    target.hotplug.fail(cpu, state)
except TargetStableError:
    return False

try:
    cls._online(target, cpu, up)
    target.hotplug.fail(cpu, -1)
    cls._online(target, cpu, not up)
except TargetStableError:
    return True
else:
    return False

```